src/Biogeochemistry.jl

```diff
@@ -0,0 +1,199 @@
+    c = fields[tracer_name]
+        
+    return src - div_Uc(i, j, k, grid, scheme, U_drift, c)
+end
+
+@inline biogeochemical_transition(i, j, k, grid, bgc, val_tracer_name, clock, fields) =
+    bgc(i, j, k, grid, val_tracer_name, clock, fields)
+
+@inline (bgc::AbstractBiogeochemistry)(i, j, k, grid, val_tracer_name, clock, fields) = zero(grid)
+
+#####
+##### Continuous form biogeochemical source
+#####
+ 
+"""Return the biogeochemical forcing for `val_tracer_name` when model is called."""
+abstract type AbstractContinuousFormBiogeochemistry <: AbstractBiogeochemistry end
+
+@inline extract_biogeochemical_fields(i, j, k, grid, fields, names::NTuple{1}) =
+    @inbounds (fields[names[1]][i, j, k],)
+
+@inline extract_biogeochemical_fields(i, j, k, grid, fields, names::NTuple{2}) =
+    @inbounds (fields[names[1]][i, j, k],
+               fields[names[2]][i, j, k])
+
+@inline extract_biogeochemical_fields(i, j, k, grid, fields, names::NTuple{N}) where N =
+    @inbounds ntuple(n -> fields[names[n]][i, j, k], Val(N))
+
+@inline function biogeochemical_transition(i, j, k, grid, bgc::AbstractContinuousFormBiogeochemistry,
+                                           val_tracer_name, clock, fields)
+
+    names_to_extract = tuple(required_biogeochemical_tracers(bgc)...,
+                             required_biogeochemical_auxiliary_fields(bgc)...)
+
+    fields_ijk = extract_biogeochemical_fields(i, j, k, grid, fields, names_to_extract)
+
+    x = xnode(Center(), Center(), Center(), i, j, k, grid)
+    y = ynode(Center(), Center(), Center(), i, j, k, grid)
+    z = znode(Center(), Center(), Center(), i, j, k, grid)
+
+    return bgc(val_tracer_name, x, y, z, clock.time, fields_ijk...)
+end
+
+@inline (bgc::AbstractContinuousFormBiogeochemistry)(val_tracer_name, x, y, z, t, fields...) = zero(x)
+
+struct NoBiogeochemistry <: AbstractBiogeochemistry end
+
+tracernames(tracers) = keys(tracers)
+tracernames(tracers::Tuple) = tracers
+
+@inline function all_fields_present(fields::NamedTuple, required_fields, grid)
+
+    for field_name in required_fields
+        if !(field_name in keys(fields))
+            fields = merge(fields, NamedTuple{(field_name, )}((CenterField(grid), )))
+        end
+    end
+
+    return fields
+end
+
+@inline all_fields_present(fields::Tuple, required_fields, grid) = (fields..., required_fields...)
+
+"""Ensure that `tracers` contains biogeochemical tracers and `auxiliary_fields` contains biogeochemical auxiliary fields (e.g. PAR)."""
+@inline function validate_biogeochemistry(tracers, auxiliary_fields, bgc, grid, clock)
+    req_tracers = required_biogeochemical_tracers(bgc)
+    tracers = all_fields_present(tracers, req_tracers, grid)
+
+    req_auxiliary_fields = required_biogeochemical_auxiliary_fields(bgc)
+
+    all(field ∈ tracernames(auxiliary_fields) for field in req_auxiliary_fields) ||
+        error("$(req_auxiliary_fields) must be among the list of auxiliary fields to use $(typeof(bgc).name.wrapper)")
```

Comment:
Do we want required auxiliary fields? I think this makes models harder to implement. It's not necessary, because any biogeochemistry-speciifc auxiliary fields can be computed in `update_biogeochemical_state!`.

Reply:
I started rewriting one of the models we've been using using the new interface and think we need required auxiliary fields to check the user is setting up PAR fields for the model.

[Heres](https://github.com/OceanBioME/OceanBioME.jl/blob/Oceananigans-Update/examples/column.jl) an example of a script using it, and [the model code](https://github.com/OceanBioME/OceanBioME.jl/tree/Oceananigans-Update/src/Models/AdvectedPopulations/LOBSTER).

Reply:
I think I understand --- is it because we currently only have access to `model.auxiliary_fields` within a biogeochemical transition function?

We can perhaps generalize this by using an extractor function to get the auxiliary fields for continuous form biogeochemistry, and merging the model auxiliary fields with the biogeochemical auxiliary fields there... what do you think?

Reply:
Those links go to a private repo (but I can see them)

Reply:
You can get auxiliary fields in normal forcing functions I think, my issue is just that if we can't check if a users put an auxiliary field in then it would probably give some weird errors when we try to integrate the PAR etc.

Do you mean to have a separate property for the models with biogeochemical auxiliary fields?

Reply:
> Those links go to a private repo (but I can see them)

Oh yeah I forgot, they will be public soon

Reply:
> Do you mean to have a separate property for the models with biogeochemical auxiliary fields?

We can implement an interface function `biogeochemical_auxiliary_fieilds(bgc)` that graps any "additional" auxiliary fields that are not part of `model.auxiliary_fields`, and merges them into the list of fields that we subsequently sift through for the purpose of passing them to the transition functions.

This would then relieve the user from needing to explicitly pass things like `PAR` into `auxiliary_fields` in the model constructor.

Reply:
Ah I see, so the auxiliary field is part of the bgc instead? I think this make sense

Reply:
Yes, basically `model.biogeochemistry` could then potentially have its own list of auxiliary fields, leaving `model.auxiliary_fields` untouched. (Perhaps `model.auxiliary_fields` would be better called `model.user_auxiliary_fields`.)